### ISSUE TITLE
Large Distributed Table Upgrades

### ISSUE THREAD

**Issue by cyxgun** [citusdata/citus#2500]
Our product uses Citus. When shipping new product versions, we will need to upgrade older schema versions to the latest schema version while preserving data.  Citus is targeted for large data pools, and we have a large distributed table that can use well over half the available disk space.   Some obvious Postgres table upgrades (UPDATE, ALTER TABLE) require free space equivalent to the table size.  So, if one wants to optimize disk usage, and use a large distributed table while maintaining data across upgrades, there is a problem.

Because Citus shards the data into smaller tables "under the hood", it is uniquely positioned to support the "large table upgrade" by upgrading each table separately, and allowing the space to be reclaimed on a per-table basis.  Some caveats may include:
- Database is not available for other transactions during the upgrade
- Rollback may be a challenge or not supported.
- Full multi-shard performance may not be available to minimize bloat.

Or, perhaps, there are other approaches (e.g. add another worker node)?

**Comment by onderkalaci** [comment]
Two suggestions/thoughts:

1) Use sequential execution mode

I don't know if the following helps you or not, but you can execute commands one shard at a time. You should do something like the following::

```SQL
BEGIN;
SET LOCAL citus.multi_shard_modify_mode TO sequential ;
ALTER TABLE ....; (or UPDATE)
COMMIT;
```

At the same time, if you set the autovacuum settings aggressive enough on the workers, it might be good enough.

2) Try to use PG 11 and fast `ALTER TABLE %s ADD COLUMN %s type DEFAULT val` if possible. See an [example](https://www.depesz.com/2018/04/04/waiting-for-postgresql-11-fast-alter-table-add-column-with-a-non-null-default/).

So, something like the following should work reasonably fast if you're adding/removing columns or updating all columns to the same value.

```SQL

BEGIN;

-- doesn't re-write table for pg > 11
ALTER TABLE products ADD COLUMN new_version text DEFAULT '1.0';

-- drop the old version column, which doesn't re-write table
ALTER TABLE products DROP COLUMN version;

-- finally rename the column which is instant
ALTER TABLE products RENAME COLUMN new_version TO version;

COMMIT;
```

**Comment by marcocitus** [comment]
> I don't know if the following helps you or not, but you can execute commands one shard at a time. You should do something like the following::

I don't think that helps because a full copy will still be made. You can however do:

```
-- Add the column to shards one at a time and commit each command immediately
SELECT run_command_on_placements('test', 'ALTER TABLE %s ADD COLUMN z int default 3', false);

-- Add the column to the distributed table without propagating to shards
BEGIN;
SET LOCAL citus.enable_ddl_propagation TO off;
ALTER TABLE test ADD COLUMN z int default 3;
END;
```
